sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -347,13 +367,58 @@ def test_32_equal_64(input_dtype, encode):
     X_input = np.array(X, dtype=input_dtype)
 
     # 32 bit output
-    kbd_32 = KBinsDiscretizer(n_bins=3, encode=encode, dtype=np.float32)
+    kbd_32 = KBinsDiscretizer(n_bins=3, encode=encode, dtype=np.float32, subsample=None)
     kbd_32.fit(X_input)
     Xt_32 = kbd_32.transform(X_input)
 
     # 64 bit output
-    kbd_64 = KBinsDiscretizer(n_bins=3, encode=encode, dtype=np.float64)
+    kbd_64 = KBinsDiscretizer(n_bins=3, encode=encode, dtype=np.float64, subsample=None)
     kbd_64.fit(X_input)
     Xt_64 = kbd_64.transform(X_input)
 
     assert_allclose_dense_sparse(Xt_32, Xt_64)
+
+
+def test_subsample_full_and_none():
+    X = np.array([-2, 1.5, -4, -1]).reshape(-1, 1)
+    kbd_subs_none = KBinsDiscretizer(
+        n_bins=10, encode="ordinal", strategy="quantile", subsample=None
+    )
+    kbd_subs_xlen = KBinsDiscretizer(
+        n_bins=10, encode="ordinal", strategy="quantile", subsample=X.shape[0]
+    )
+    kbd_subs_none.fit(X)
+    kbd_subs_xlen.fit(X)
+
+    assert_array_equal(kbd_subs_none.bin_edges_[0], kbd_subs_xlen.bin_edges_[0])
+    assert kbd_subs_none.bin_edges_.shape == kbd_subs_xlen.bin_edges_.shape
+
+
+def test_subsample_default_and_none():
+    X = np.array([-2, 1.5, -4, -1]).reshape(-1, 1)
+    kbd_subs_none = KBinsDiscretizer(
+        n_bins=10, encode="ordinal", strategy="quantile", subsample=None
+    )
+    kbd_subs_def = KBinsDiscretizer(n_bins=10, encode="ordinal", strategy="quantile")
+    kbd_subs_none.fit(X)
+
+    msg = "Pass subsample=None"
+    with pytest.warns(FutureWarning, match=msg):
+        kbd_subs_def.fit(X)
+
+    assert_array_equal(kbd_subs_none.bin_edges_[0], kbd_subs_def.bin_edges_[0])
+    assert kbd_subs_none.bin_edges_.shape == kbd_subs_def.bin_edges_.shape
```

Comment:
We can parametrize these 2 tests:

```suggestion
# FIXME: we should adapt this test since the default will not be equivalent
# anymore to subsample=None or subsample=X.shape[0]. Also, not FutureWarning
# will be raised anymore.
@pytest.mark.parametrize("subsample", [None, "full"])
def test_kbinsdiscretizer_subsample(subsample):
    """Check the consistency of the default strategy of `subsample`."""
    X = np.array([-2, 1.5, -4, -1]).reshape(-1, 1)
    subsample = X.shape[0] if subsample == "full" else subsample

    kbd_default = KBinsDiscretizer(n_bins=10, encode="ordinal", strategy="quantile")

    msg = "In version 1.2, onwards subsample = 2e5 will be used by default. "
    with pytest.warns(FutureWarning, match=msg):
        kbd_default.fit(X)

    kbd_with_subsampling = clone(kbd_default)
    kbd_with_subsampling.set_params(subsample=subsample).fit(X)

    for bin_kbd_default, bin_kbd_with_subsampling in (
        zip(kbd_default.bin_edges_[0], kbd_with_subsampling.bin_edges_[0])
    ):
        np.testing.assert_allclose(bin_kbd_default, bin_kbd_with_subsampling)

    assert kbd_default.bin_edges_.shape == kbd_with_subsampling.bin_edges_.shape
```